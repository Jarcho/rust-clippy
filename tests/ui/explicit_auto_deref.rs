// run-rustfix

#![allow(
    unused_imports,
    dead_code,
    clippy::borrowed_box,
    clippy::deref_addrof,
    clippy::useless_vec
)]
#![warn(clippy::explicit_auto_deref)]

use std::{
    ops::{Deref, DerefMut},
    path::{Path, PathBuf},
};

fn main() {
    let _: &str = &*String::new();
    let _: &Path = &*PathBuf::new();
    let _: &[_] = vec![0u32, 1u32, 2u32].deref();

    fn f2<T>(_: &[T]) {}
    f2(vec![0u32].deref());

    fn f3() -> &'static str {
        static S: String = String::new();
        &*S
    }

    fn f4<T1, T2>(_: &(T1, T2)) {}
    f4(Box::new((0u32, 0i32)).deref());

    let f5 = |_: &str| {};
    f5(&*String::new());

    fn f6(f: impl Fn(&str)) {
        f(&*String::new());
    }

    fn f7(f: &dyn Fn(&str)) {
        f(&*String::new());
    }

    let _: &Box<_> = &*Box::new(Box::new(0));
    let _: Box<u32> = *Box::new(Box::new(0));

    fn f8<T: ?Sized>(_: &T) {}
    f8(&*String::new());

    struct S1<T>(T);
    impl<T: Fn(&str)> S1<T> {
        fn f(&self, f: impl Fn(&str)) {
            f(&*String::new());
            (self.0)(&*String::new());
        }
    }

    fn f9<T>(f: &mut T)
    where
        T: Iterator,
        <T as Iterator>::Item: Fn(&str),
    {
        f.next().unwrap()(&*String::new())
    }

    struct S2<'a> {
        x: &'a str,
    }
    let _ = S2 { x: &*String::new() };

    struct S3<'a>(&'a str);
    let _ = S3(&*String::new());

    macro_rules! m1 {
        ($e:expr) => {{
            fn f(_: &str) {}
            f($e);
        }};
    }
    m1!(&*String::new());

    macro_rules! m2 {
        ($e:expr) => {
            &$e
        };
    }
    let _: &str = &**m2!(String::new());
}
