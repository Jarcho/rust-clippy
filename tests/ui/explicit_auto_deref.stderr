error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:18:20
   |
LL |     let _: &str = &*String::new();
   |                    ^^^^^^^^^^^^^^ help: try this: `String::new()`
   |
   = note: `-D clippy::explicit-auto-deref` implied by `-D warnings`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:19:21
   |
LL |     let _: &Path = &*PathBuf::new();
   |                     ^^^^^^^^^^^^^^^ help: try this: `PathBuf::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:20:19
   |
LL |     let _: &[_] = vec![0u32, 1u32, 2u32].deref();
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `&vec![0u32, 1u32, 2u32]`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:23:8
   |
LL |     f2(vec![0u32].deref());
   |        ^^^^^^^^^^^^^^^^^^ help: try this: `&vec![0u32]`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:27:10
   |
LL |         &*S
   |          ^^ help: try this: `S`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:31:8
   |
LL |     f4(Box::new((0u32, 0i32)).deref());
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `&Box::new((0u32, 0i32))`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:34:9
   |
LL |     f5(&*String::new());
   |         ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:37:12
   |
LL |         f(&*String::new());
   |            ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:41:12
   |
LL |         f(&*String::new());
   |            ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:53:16
   |
LL |             f(&*String::new());
   |                ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:54:23
   |
LL |             (self.0)(&*String::new());
   |                       ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:63:28
   |
LL |         f.next().unwrap()(&*String::new())
   |                            ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:69:22
   |
LL |     let _ = S2 { x: &*String::new() };
   |                      ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:72:17
   |
LL |     let _ = S3(&*String::new());
   |                 ^^^^^^^^^^^^^^ help: try this: `String::new()`

error: explicit dereference in auto-deref context
  --> $DIR/explicit_auto_deref.rs:87:19
   |
LL |     let _: &str = &**m2!(String::new());
   |                   ^^^^^^^^^^^^^^^^^^^^^ help: try this: `m2!(String::new())`

error: aborting due to 15 previous errors

